packages/core/src/new-api/internal/execution/execution-engine.ts

```diff
@@ -0,0 +1,219 @@
+import { IgnitionError } from "../../../errors";
+import { isDeploymentExecutionState } from "../../../internal/utils/guards";
+import { isRuntimeValue } from "../../type-guards";
+import { DeploymentResult } from "../../types/deployer";
+import { ExecutionEngineState } from "../../types/execution-engine";
+import {
+  ExecutionStateMap,
+  ExecutionStatus,
+} from "../../types/execution-state";
+import {
+  ExecutionResult,
+  FutureRestart,
+  FutureStart,
+  JournalableMessage,
+} from "../../types/journal";
+import {
+  AccountRuntimeValue,
+  Future,
+  FutureType,
+  RuntimeValueType,
+} from "../../types/module";
+import { safeToString } from "../reconciliation/utils";
+
+import { executionStateReducer } from "./executionStateReducer";
+import { isExecutionResult, isOnChainAction, isOnchainResult } from "./guards";
+
+type ExecutionBatch = Future[];
+
+export class ExecutionEngine {
+  public async execute(state: ExecutionEngineState): Promise<DeploymentResult> {
+    const { batches, module } = state;
+
+    const futures = module.getFutures();
+
+    for (const batch of batches) {
+      // TODO: consider changing batcher to return futures rather than ids
+      const executionBatch = batch.map((futureId) =>
+        this._lookupFuture(futures, futureId)
+      );
+
+      const batchResult = await this._executeBatch(executionBatch, state);
+
+      if (batchResult.some((b) => b.type === "execution-failure")) {
+        return { status: "failed" };
+      }
+
+      if (batchResult.some((b) => b.type === "execution-hold")) {
+        return { status: "hold" };
+      }
+
+      if (batchResult.every((b) => b.type !== "execution-success")) {
+        throw new IgnitionError("Unexpected state");
+      }
+    }
+
+    return {
+      status: "success",
+      contracts: this._resolveDeployedContractsFrom(state),
+    };
+  }
+
+  private async _executeBatch(
+    batch: ExecutionBatch,
+    state: ExecutionEngineState
+  ) {
+    return Promise.all(
+      batch.map((entry) => this._executeBatchEntry(entry, state))
+    );
+  }
+
+  private async _executeBatchEntry(
+    future: Future,
+    state: ExecutionEngineState
+  ): Promise<ExecutionResult> {
+    let current: JournalableMessage = this._startOrRestartFor(future, state);
+    await this._apply(state, current);
+
+    const context = {
+      executionState: state.executionStateMap[future.id],
+    };
+
+    const exectionStrategy = state.strategy.executeStrategy(context);
+
+    while (!isExecutionResult(current)) {
+      context.executionState = state.executionStateMap[future.id];
+
+      if (isOnChainAction(current)) {
+        current = await state.transactionService.onchain(current);
```

Comment:
I think this is missing the fact that the transaction service will need to do things that must be journaled internally (e.g. gas bump). Have you thought about it? Should it receive the entire state to be able to mutate it?

Reply:
Or maybe I'm missing how the transaction service should work